runge_kutta.py

```diff
@@ -0,0 +1,228 @@
+                                 xyz[1] + (1 / 4 + np.sqrt(3) / 6) * k_coeffs[2] + k_coeffs[3] / 4,
+                                 xyz[2] + (1 / 4 + np.sqrt(3) / 6) * k_coeffs[4] + k_coeffs[5] / 4),
+        # equation for k_coeffs[5] k2z
+        k_coeffs[5] - h * func_z(xyz[0] + (1 / 4 + np.sqrt(3) / 6) * k_coeffs[0] + k_coeffs[1] / 4,
+                                 xyz[1] + (1 / 4 + np.sqrt(3) / 6) * k_coeffs[2] + k_coeffs[3] / 4,
+                                 xyz[2] + (1 / 4 + np.sqrt(3) / 6) * k_coeffs[4] + k_coeffs[5] / 4),
+    ]
+    return coeff_system
+
+
+def solve_ksystem(system_to_solve: callable, xyz: tuple, h: float, k0: np.array = np.array((0, 0, 0, 0, 0, 0))) -> np.array:
+    """
+
+    :param system_to_solve: callable function that returns list of equations to solve
+    :param k0: initial values for optimizing variables
+    :param hxyz: initial conditions for the system
+    :return: list of roots (k_coeffs)
+    """
+    hxyz = (h, *xyz)
+    root = fsolve(system_to_solve, k0, args=hxyz)
+    return root
+
+
+def make_one_step(xyz, h):
+    """"
+    Makes one step with the step size = h
+    """
+    k_coeffs = solve_ksystem(system_to_solve=ksystem, xyz=xyz, h=h)
+    xyz_one_step = np.array([xyz[0] + (k_coeffs[0] + k_coeffs[2]) / 2,
+                             xyz[1] + (k_coeffs[2] + k_coeffs[3]) / 2,
+                             xyz[2] + (k_coeffs[4] + k_coeffs[5]) / 2
+                             ])
+    return xyz_one_step
+
+
+def make_two_steps(xyz: tuple, h: float):
+    """"
+    Makes two steps with the step size = h/2
+    """
+    # k system coeffs (roots) for step = h/2
+    h = h / 2
+    k_coeffs = solve_ksystem(system_to_solve=ksystem, xyz=xyz, h=h)
+
+    # values for xyz with one step = h/2
+    xyz_one_half_step = np.array([xyz[0] + (k_coeffs[0] + k_coeffs[2]) / 2,
+                                  xyz[1] + (k_coeffs[2] + k_coeffs[3]) / 2,
+                                  xyz[2] + (k_coeffs[4] + k_coeffs[5]) / 2
+                                  ])
+
+    # values for xyz with two steps of step = h/2
+    xyz_two_half_step = np.array([xyz_one_half_step[0] + (k_coeffs[0] + k_coeffs[2]) / 2,
+                                  xyz_one_half_step[1] + (k_coeffs[2] + k_coeffs[3]) / 2,
+                                  xyz_one_half_step[2] + (k_coeffs[4] + k_coeffs[5]) / 2
+                                  ])
+    return xyz_two_half_step
+
+
+def calculate_iter_errors(xyz_one_step: np.array, xyz_two_half_step: np.array) -> np.array:
+    """"
+    Calculates iteration errors for each coordinate (x,y,z)
+    """
+    iter_errors = np.array([np.abs((xyz_one_step[0] - xyz_two_half_step[0]) / 15),
+                            np.abs((xyz_one_step[1] - xyz_two_half_step[1]) / 15),
+                            np.abs((xyz_one_step[2] - xyz_two_half_step[2]) / 15)
+                            ])
+    return iter_errors
+
+
+def euclidian_norm(arr: np.array):
```

Comment:
check out `numpy.linalg.norm`

Reply:
@aartiukh it's not euclidian but The Frobenius norm

Reply:
they are equivalent

Reply:
ok done